File: idexal-ide/idexa-cli/src/utils/project.ts
import fs from 'fs-extra';
import path from 'path';

export interface ProjectInfo {
  name: string;
  type: string;
  rootPath: string;
  fileCount: number;
  languages: string[];
  primaryLanguage: string;
  dependencies: string[];
}

const PROJECT_INDICATORS: Record<string, { files: string[]; language: string }> = {
  javascript: { files: ['package.json', 'yarn.lock', 'pnpm-lock.yaml'], language: 'javascript' },
  typescript: { files: ['tsconfig.json'], language: 'typescript' },
  react: { files: ['react.config.js', '.babelrc'], language: 'typescript' },
  nextjs: { files: ['next.config.js', 'next.config.mjs'], language: 'typescript' },
  python: { files: ['requirements.txt', 'setup.py', 'pyproject.toml', 'Pipfile'], language: 'python' },
  rust: { files: ['Cargo.toml'], language: 'rust' },
  go: { files: ['go.mod'], language: 'go' },
  java: { files: ['pom.xml', 'build.gradle'], language: 'java' },
  ruby: { files: ['Gemfile'], language: 'ruby' },
  php: { files: ['composer.json'], language: 'php' }
};

export async function detectProject(searchPath?: string): Promise<ProjectInfo | null> {
  const rootPath = searchPath || process.cwd();
  
  const type = await detectProjectType(rootPath);
  if (!type) return null;

  const name = await getProjectName(rootPath, type);
  const fileCount = await countFiles(rootPath);
  const languages = await detectLanguages(rootPath, type);
  const dependencies = await getDependencies(rootPath, type);

  return {
    name,
    type,
    rootPath,
    fileCount,
    languages,
    primaryLanguage: PROJECT_INDICATORS[type]?.language || 'unknown',
    dependencies
  };
}

async function detectProjectType(rootPath: string): Promise<string | null> {
  for (const [type, { files }] of Object.entries(PROJECT_INDICATORS)) {
    for (const file of files) {
      if (await fs.pathExists(path.join(rootPath, file))) {
        return type;
      }
    }
  }
  return null;
}

async function getProjectName(rootPath: string, type: string): Promise<string> {
  try {
    if (type === 'javascript' || type === 'typescript' || type === 'react' || type === 'nextjs') {
      const pkg = await fs.readJSON(path.join(rootPath, 'package.json'));
      return pkg.name || path.basename(rootPath);
    }
    if (type === 'rust') {
      const cargo = await fs.readFile(path.join(rootPath, 'Cargo.toml'), 'utf-8');
      const nameMatch = cargo.match(/name\s*=\s*"([^"]+)"/);
      return nameMatch?.[1] || path.basename(rootPath);
    }
    if (type === 'go') {
      const goMod = await fs.readFile(path.join(rootPath, 'go.mod'), 'utf-8');
      const nameMatch = goMod.match(/module\s+(.+)/);
      return nameMatch?.[1]?.split('/').pop() || path.basename(rootPath);
    }
  } catch {}
  return path.basename(rootPath);
}

async function countFiles(rootPath: string): Promise<number> {
  let count = 0;
  const ignoreDirs = ['node_modules', '.git', 'dist', 'build', '__pycache__'];
  
  const countDir = async (dir: string) => {
    const entries = await fs.readdir(dir, { withFileTypes: true });
    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (!ignoreDirs.includes(entry.name)) {
          await countDir(path.join(dir, entry.name));
        }
      } else {
        count++;
      }
    }
  };
  
  try {
    await countDir(rootPath);
  } catch {}
  
  return count;
}

async function detectLanguages(rootPath: string, type: string): Promise<string[]> {
  const languages: Set<string> = new Set();
  
  if (['javascript', 'typescript', 'react', 'nextjs'].includes(type)) {
    languages.add('JavaScript/TypeScript');
  }
  if (type === 'python') languages.add('Python');
  if (type === 'rust') languages.add('Rust');
  if (type === 'go') languages.add('Go');
  if (type === 'java') languages.add('Java');
  if (type === 'ruby') languages.add('Ruby');
  if (type === 'php') languages.add('PHP');
  
  return Array.from(languages);
}

async function getDependencies(rootPath: string, type: string): Promise<string[]> {
  try {
    if (['javascript', 'typescript', 'react', 'nextjs'].includes(type)) {
      const pkg = await fs.readJSON(path.join(rootPath, 'package.json'));
      return Object.keys({
        ...pkg.dependencies,
        ...pkg.devDependencies
      });
    }
    if (type === 'python') {
      if (await fs.pathExists(path.join(rootPath, 'requirements.txt'))) {
        const content = await fs.readFile(path.join(rootPath, 'requirements.txt'), 'utf-8');
        return content.split('\n').filter(l => l.trim() && !l.startsWith('#')).map(l => l.split('==')[0].split('>=')[0].trim());
      }
    }
    if (type === 'rust') {
      const cargo = await fs.readFile(path.join(rootPath, 'Cargo.toml'), 'utf-8');
      const deps = cargo.match(/\[dependencies\]([\s\S]*?)(?=\[|$)/);
      return deps?.[1]?.match(/(\w+)\s*=/g)?.map(d => d.replace('=', '').trim()) || [];
    }
  } catch {}
  return [];
}

File: idexal-ide/idexal-mcp-server/src/tools/project.rs
use rmcp::model::Parameters;
use serde::{Deserialize, Serialize};
use schemars::JsonSchema;
use std::path::Path;
use walkdir::WalkDir;

#[derive(Debug, Deserialize, JsonSchema)]
pub struct ProjectInfoParams {
    pub directory: Option<String>,
}

#[derive(Debug, Deserialize, JsonSchema)]
pub struct DependenciesParams {
    pub directory: Option<String>,
}

pub async fn get_project_info(params: Parameters<ProjectInfoParams>) -> Result<String, String> {
    let p = params.inner();
    let dir = p.directory.unwrap_or_else(|| ".".to_string());
    
    let mut info = serde_json::json!({
        "directory": dir,
        "files": {},
        "languages": {},
        "total_files": 0,
        "total_size": 0
    });
    
    let mut total_files = 0;
    let mut total_size = 0u64;
    let mut languages: std::collections::HashMap<String, (usize, u64)> = std::collections::HashMap::new();
    
    for entry in WalkDir::new(&dir)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| e.file_type().is_file())
    {
        // Skip hidden files and common non-source directories
        let path = entry.path().to_string_lossy();
        if path.contains("/.") || path.contains("/node_modules/") || path.contains("/target/") {
            continue;
        }
        
        let name = entry.file_name().to_string_lossy().to_string();
        let ext = Path::new(&name)
            .extension()
            .map(|e| e.to_string_lossy().to_string())
            .unwrap_or_else(|| "unknown".to_string());
        
        let metadata = entry.metadata().ok();
        let size = metadata.as_ref().map(|m| m.len()).unwrap_or(0);
        
        total_files += 1;
        total_size += size;
        
        let lang = get_language_from_ext(&ext);
        let entry = languages.entry(lang).or_insert((0, 0));
        entry.0 += 1;
        entry.1 += size;
    }
    
    // Convert to serializable format
    let langs: std::collections::HashMap<String, serde_json::Value> = languages.iter()
        .map(|(lang, (count, size))| {
            (lang.clone(), serde_json::json!({
                "files": count,
                "size": size,
                "size_formatted": format_size(*size)
            }))
        })
        .collect();
    
    info["languages"] = serde_json::json!(langs);
    info["total_files"] = serde_json::json!(total_files);
    info["total_size"] = serde_json::json!(total_size);
    info["total_size_formatted"] = serde_json::json!(format_size(total_size));
    
    // Check for common project files
    let mut project_files = Vec::new();
    if Path::new("Cargo.toml").exists() { project_files.push("Cargo.toml (Rust)"); }
    if Path::new("package.json").exists() { project_files.push("package.json (Node.js)"); }
    if Path::new("pyproject.toml").exists() { project_files.push("pyproject.toml (Python)"); }
    if Path::new("go.mod").exists() { project_files.push("go.mod (Go)"); }
    if Path::new("pom.xml").exists() { project_files.push("pom.xml (Java)"); }
    
    info["project_files"] = serde_json::json!(project_files);
    
    Ok(serde_json::to_string_pretty(&info).unwrap_or_default())
}

pub async fn list_dependencies(params: Parameters<DependenciesParams>) -> Result<String, String> {
    let p = params.inner();
    let dir = p.directory.unwrap_or_else(|| ".".to_string());
    
    let mut dependencies = Vec::new();
    
    // Check for Cargo.toml (Rust)
    let cargo_path = Path::new(&dir).join("Cargo.toml");
    if cargo_path.exists() {
        let content = std::fs::read_to_string(&cargo_path)
            .map_err(|e| format!("Failed to read Cargo.toml: {}", e))?;
        
        // Simple parsing - look for [dependencies] section
        let mut in_deps = false;
        for line in content.lines() {
            if line.trim() == "[dependencies]" {
                in_deps = true;
                continue;
            }
            if line.starts_with('[') && in_deps {
                in_deps = false;
                continue;
            }
            if in_deps && !line.trim().is_empty() && !line.trim().starts_with('#') {
                if let Some(name) = line.split('=').next() {
                    let name = name.trim().to_string();
                    dependencies.push(serde_json::json!({
                        "name": name,
                        "type": "rust",
                        "source": "cargo"
                    }));
                }
            }
        }
    }
    
    // Check for package.json (Node.js)
    let package_path = Path::new(&dir).join("package.json");
    if package_path.exists() {
        let content = std::fs::read_to_string(&package_path)
            .map_err(|e| format!("Failed to read package.json: {}", e))?;
        
        if let Ok(pkg) = serde_json::from_str::<serde_json::Value>(&content) {
            if let Some(deps) = pkg["dependencies"].as_object() {
                for name in deps.keys() {
                    dependencies.push(serde_json::json!({
                        "name": name,
                        "type": "node",
                        "source": "npm"
                    }));
                }
            }
            if let Some(dev_deps) = pkg["devDependencies"].as_object() {
                for name in dev_deps.keys() {
                    dependencies.push(serde_json::json!({
                        "name": name,
                        "type": "node",
                        "source": "npm",
                        "dev": true
                    }));
                }
            }
        }
    }
    
    // Check for requirements.txt or pyproject.toml (Python)
    let requirements_path = Path::new(&dir).join("requirements.txt");
    if requirements_path.exists() {
        let content = std::fs::read_to_string(&requirements_path)
            .map_err(|e| format!("Failed to read requirements.txt: {}", e))?;
        
        for line in content.lines() {
            let line = line.trim();
            if !line.is_empty() && !line.starts_with('#') && !line.starts_with('-') {
                let name = line.split(|c: char| c == '=' || c == '>' || c == '<' || c == '!')
                    .next()
                    .unwrap_or(line)
                    .trim()
                    .to_string();
                dependencies.push(serde_json::json!({
                    "name": name,
                    "type": "python",
                    "source": "pip"
                }));
            }
        }
    }
    
    let result = serde_json::json!({
        "directory": dir,
        "total_dependencies": dependencies.len(),
        "dependencies": dependencies
    });
    
    Ok(serde_json::to_string_pretty(&result).unwrap_or_default())
}

fn get_language_from_ext(ext: &str) -> String {
    match ext {
        "rs" => "Rust".to_string(),
        "ts" | "tsx" => "TypeScript".to_string(),
        "js" | "jsx" => "JavaScript".to_string(),
        "py" => "Python".to_string(),
        "go" => "Go".to_string(),
        "java" => "Java".to_string(),
        "rb" => "Ruby".to_string(),
        "cpp" | "cc" | "cxx" => "C++".to_string(),
        "c" | "h" => "C".to_string(),
        "cs" => "C#".to_string(),
        "swift" => "Swift".to_string(),
        "kt" => "Kotlin".to_string(),
        "vue" => "Vue".to_string(),
        "svelte" => "Svelte".to_string(),
        "html" | "htm" => "HTML".to_string(),
        "css" | "scss" | "less" => "CSS".to_string(),
        "json" => "JSON".to_string(),
        "yaml" | "yml" => "YAML".to_string(),
        "toml" => "TOML".to_string(),
        "md" | "markdown" => "Markdown".to_string(),
        "sql" => "SQL".to_string(),
        "sh" | "bash" => "Shell".to_string(),
        _ => "Other".to_string(),
    }
}

fn format_size(bytes: u64) -> String {
    const KB: u64 = 1024;
    const MB: u64 = 1024 * KB;
    const GB: u64 = 1024 * MB;
    
    if bytes >= GB {
        format!("{:.2} GB", bytes as f64 / GB as f64)
    } else if bytes >= MB {
        format!("{:.2} MB", bytes as f64 / MB as f64)
    } else if bytes >= KB {
        format!("{:.2} KB", bytes as f64 / KB as f64)
    } else {
        format!("{} B", bytes)
    }
}